### DISCUSSION TITLE
what info are encoded to form the nodeid and used in gotonode in URL parameter.

### DISCUSSION THREAD

**Discussion by a-imran99** [Ylianst/MeshCentral#4062] [General]
I just wanted to use the gotonode in url to directly go to the specific device or host. I wanted to know what information about the device is encoded to form the nodeid and used in meshcentral.

We are using meshcentral in Iframe and wanted to connect to the specific device from URL.

need help.!

**Comment by Ylianst** [comment]
Hi. Excellent question. The NodeID is the base64 encoded SHA384 hash of the device's certificate public key. The resulting base64 string is modified a bit to work in a URL, 2 of the base64 characters are replaced. This is a great NodeID for security since this identifier is "provable", when a node connects to MeshCentral, in the authentication exchange, MeshCentral can compute the node's identifier from the presented certificate so, a connection can't be spoofed.

You can get the list of NodeID's from the database or using MeshCtrl.js like this:

```
node node_modules/meshcenral/meshctrl.js listdevices
```

You get something list this:

![image](https://user-images.githubusercontent.com/1319013/171458468-3a5623d3-ffef-4054-9d6d-4fa7eb620303.png)

You can also add `--json` to get the result in JSON format. Hopefully this helps.

**Reply by a-imran99** [reply]
Hi there, thanks for taking time to reply to the discussion. The above reply was really helpful

now I get a clear idea of how a node Id is formed and further reading into the documentation I came to know that the node id is created while installing the mesh agent into the host system and the node id is the base64 encoded of the SHA384 hash and this node Id changes every time we uninstall and reinstall the mesh agent. So I guess the node id is not same all the time for a single machine. So we decided to store the node id of all the devices in which the mesh agent is installed and then use them to connect to a particular device when needed. All this because we intend to use meshcentral inside Iframe :) 

So I have another imported question and sorry for taking this discussion to another topic. 
 
Can we bypass the login process for meshcentral because the login page for the parent application already exist and I don't want the user to login again into mesh central. Is there anyway I can achieve that. 

It would be great if you reply. Thank you and you are really cool :) .

**Comment by Ylianst** [comment]
To finish off on the first topic, you may want to take a look at agent tags, [I have a video on this here](https://www.youtube.com/watch?v=KPMtzzLWoK0). This allows you to create your own unique identifiers for your devices. Also you can `gotodevicename` like this if you like:

```
https://central.mesh.meshcentral.com/?viewmode=10&gotodevicename=central
```

But this will only work right if the device names are unique.

For your next question on login, yes, look at [section 14 of the MeshCentral User's guide](https://meshcentral.com/info/docs/MeshCentral2UserGuide.pdf). This is the best system. Your main server needs to generate a login token using a private key that it shares with the MeshCentral server. You can then pass this in the URL to login. Let me know if that works.

**Reply by a-imran99** [reply]
Thanks for the reply, will try to implement the above. Thanks again.

**Comment by brettgilmer** [comment]
I am doing something similar enough that commenting on this post makes sense.

I am trying to programmatically convert the node returned on from the target desktop (using <service.exe> -nodeID) and convert it into the nodeID used in the server's database and for direct connectons.

The PC returns `D4F0AD096AB6B710DFFBC314989CD64824BF5DC2956257AEA485E757E6FCD7D780954DA1B596F3B9062C3AE89DD57C32`
and the meshcentral database says:
`1PCtCWq2txDf@8MUmJzWSCS$XcKVYleupIXnV@b819eAlU2htZbzuQYsOuid1Xwy`

What I have tried (in python) is:

original=b'D4F0AD096AB6B710DFFBC314989CD64824BF5DC2956257AEA485E757E6FCD7D780954DA1B596F3B9062C3AE89DD57C32'
gfg = hashlib.sha384()
gfg.update(original)
digest = gfg.digest()
print(b64encode(digest))

I have also tried converting "original" to hex first, still doesn't match
I understand that after this I need to escape-out (or sub-out) some characters but what is printed above is not even close.

getting 
qpOhtcL6kaxmDb5feAQPv1l90qYizd4VWtpaiNpDSRp75T0Sp3DkDGmMQQBc+x2w (as done above)
6Lczi5jMlYmTlf40EPF2PRBH9wggB7fr1m83jWzsmt+IOxI/g4nGLjsPj7ZDEJGX (w hex conversion)
Any ideas?
Thanks!

**Comment by krayon007** [comment]
The value on the agent side, where you passed -nodeid, I'm pretty sure is just a hex encoding of the hash, not base64.

I think you need to decode the hex first, because I think the base64 encoding is done on the raw hash not the hex encoding of it.

**Reply by brettgilmer** [reply]
awesome, thanks, the -nodeID is in fact the hex encoding of the hash.
So

original='D4F0AD096AB6B710DFFBC314989CD64824BF5DC2956257AEA485E757E6FCD7D780954DA1B596F3B9062C3AE89DD57C32'
node_id = bytes.fromhex(original)
print(b64encode(node_id))

Is correct.
